### DISCUSSION TITLE
Why are imagepolicies tied to images?

### DISCUSSION THREAD

**Discussion by Frederik-Baetens** [fluxcd/flux2#1372] [Proposals] [answered]
Why are imagepolicies tied to images? Why not either derive the image from the "image:" field like in fluxV1, or make the image update annotation contain both the imagepolicy and the image?

The way it is now is annoying because this forces me to create separate, but basically duplicate imagepolicy files for 20 of my microservices which require a separate image.

**Comment by squaremo** [comment, marked as answer]
> Why not either derive the image from the "image:" field like in fluxV1, or make the image update annotation contain both the imagepolicy and the image?

The surface answer is that the image-reflector-controller does not have access to the files in git, so it can't see what's in field values or update markers. Which leads in two directions:

*Why doesn't the image-reflector-controller work from git?*

Separating the reflection and the automation jobs is a design decision, and not inevitable. Why did I do it this way?

One reason is to make the image reflector stand on its own. You might want to use the information it gathers for some purpose other than input to the image automation controller. Requiring a git repository makes it much less generally useful.

Another reason is that in GitOps Toolkit, everything is driven through the Kubernetes API, so at least at the base level, the specification for what to scan needs to be represented as API objects.

*Can something else examine the files and make objects for image-reflector-controller?*

Yes, possibly. For example, the automation controller could examine the git repository under the path it's given, and create `ImageRepository` and `ImagePolicy` objects according to some scheme. The sticking point would likely be how to assign secrets -- these may be supplied out of band, and not necessarily in the same namespace or cluster as the objects in git.

I considered this while I was designing the image automation, and didn't take it further because 1. it was not the simplest design and 2. I would need the image reflection API, whatever else I did. But it can be revisited.

> this forces me to create separate, but basically duplicate imagepolicy files for 20 of my microservices which require a separate image.

This is not the situation for everyone, and may not even be situation for you all the time. For example, you may want to pin a particular microservice to a version while you debug it, or skip a version that's a known bad for a particular image. If all your images followed the same policy, you wouldn't be able to make exceptions like this. You could "fork" a policy when this happens, it's true.

**In summary**, the existing design is the simplest design that worked, and not the final word. It's possible to build a more sophisticated mechanism on the primitives -- though of course, the more sophisticated, the more there is to argue about :-)

**Reply by Frederik-Baetens** [reply]
Thanks for the explanation, I understand that this is a clean and versatile primitive. I am excited that the possibility of generating the imagepolicy objects is left open.

**Reply by squaremo** [reply]
(I've marked my answer as accepted, to drop this discussion below the fold; hope that's OK)

**Comment by coperator** [comment]
My solution to this is to create ImageRepositories and ImagePolicies via the same Helm chart I also use to deploy my services. So, I only need to declare the image repository once in the HelmRelease. Downside is, that ImageRepositories, ImagePolicies and the ImageUpdateAutomation with all necessary secrets must reside in the same namespace and not in flux-system.
@squaremo Do you think, this is a valid/good approach?

**Reply by Frederik-Baetens** [reply]
I've personally solved this problem by putting the imagerepositories & imagepolicies in a kustomize base, where I make use of the mostly undocumented "configurations" option in my kustomization.yaml. this allows me to add a suffix to the image in the imagerepositories, and have the imagepolicy correctly refer to the imagerepository by registering the reference in the imagepolicy to the imagerepository as a namereference.

That said, helm is a more versatile solution here, especially if you're already using helm. Kustomize is something that I've only been able to make work because of conventions we have around our image names.
